sphinxcontrib/chapeldomain.py

```diff
@@ -572,6 +573,110 @@ class information, if relevant.
         return title, target
 
 
+class ChapelModuleIndex(Index):
+    """Provides Chapel module index based on chpl:module."""
+
+    name = 'modindex'
+    localname = l_('Chapel Module Index')
+    shortname = l_('modules')
+
+    def generate(self, docnames=None):
+        """Returns entries for index given by ``name``. If ``docnames`` is given,
+        restrict to entries referring to these docnames.
+
+        Retunrs tuple of ``(content, collapse)``. ``collapse`` is bool. When
+        True, sub-entries should start collapsed for output formats that
+        support collapsing.
+
+        ``content`` is a sequence of ``(letter, entries)`` tuples. ``letter``
+        is the "heading" for the given ``entries``, in this case the starting
+        letter.
+
+        ``entries`` is a sequence of single entries, where a single entry is a
+        sequence ``[name, subtype, docname, anchor, extra, qualifier,
+        description]``. These items are:
+
+        * ``name`` - name of the index entry to be displayed
+        * ``subtype`` - sub-entry related type:
+
+          * 0 - normal entry
+          * 1 - entry with sub-entries
+          * 2 - sub-entry
+
+        * ``docname`` - docname where the entry is located
+        * ``anchor`` - anchor for the entry within docname
+        * ``extra`` - extra info for the entry
+        * ``qualifier`` - qualifier for the description
+        * ``description`` - description for the entry
+
+        Qualifier and description are not rendered in some output formats.
+        """
+        content = {}
+
+        # list of prefixes to ignore
+        ignores = self.domain.env.config['chapeldomain_modindex_common_prefix']
+        ignores = sorted(ignores, key=len, reverse=True)
+
+        # list of all modules, sorted by module name
+        modules = sorted(iteritems(self.domain.data['modules']),
+                         key=lambda x: x[0].lower())
+
+        # sort out collapsible modules
+        prev_modname = ''
+        num_toplevels = 0
+        for modname, (docname, synopsis, platforms, deprecated) in modules:
+            # If given a list of docnames and current docname is not in it,
+            # skip this docname for the index.
+            if docnames and docname not in docnames:
+                continue
+
+            for ignore in ignores:
+                if modname.startswith(ignore):
+                    modname = modname[len(ignore):]
+                    stripped = ignore
+                    break
+            else:
```

Comment:
for .. else is actually a python feature, although admittedly not an intuitive one. The else block is executed if the for loop exits normally (i.e. without calling break). So, in this case stripped is set to `''` iff the module name is not in the ignores list. 

http://psung.blogspot.com/2007/12/for-else-in-python.html